### PR TITLE
Create load data and add data methods

### PR DESCRIPTION
**Pull request by tanveerisonline**
**Thanks a lot for Reviewing our pull request for review saving data of the OOP School Library in JSON files using ruby.**

**In this PR we have done the following things**

- [ ] Create load data methods.
- [ ] Add save_ methods.
- [ ] Reading data.

### CODE REVIEW

File: data_controller.rb [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+    if File.exist?(file) && !File.empty?(file)
+      JSON.parse(File.read(file)).map(&block)
+    else
+      []
+    end
+  end
+
+  def save_data(file, data)
+    File.write(file, JSON.generate(data))
+  end
```
Comment (elisha2kyakpo1) [inline data_controller.rb]: - [ ] The app works well, I was able to create a person, book, and rental, but looking at the requirements for this project, data wasn't saved. After exiting the app using option 7, data should be saved and then loaded when the app is started. I have attached an image for your reference. You may as well fix the error that comes when a user use option 7 (see attached image).

save

**Review by elisha2kyakpo1 — CHANGES_REQUESTED** [review]
Hi 👋 @tanveerisonline and @elarhadu, 

Good job so far!
There are some issues that you still need to work on to go to the next project but you are almost there!

## Highlights

✔️ The PR title and description look good and professional.
✔️ No linter errors.
✔️ Descriptive readme file.

## Required Changes ♻️
_Check the comments under the review._

- [ ] Kindly add a correct link for the license in the readme file, currently, the license link is broken and it throws a 404 error. This way the readme file will look more professional.

## Optional suggestions
_Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you consider them as they can improve your code._

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

File: books.json [hunk 2]
```diff
@@ -0,0 +1 @@
+[{"title":"Gatsby","author":"Fitzgerald"}]
```
Comment (BenMukebo) [inline books.json]: - [ ] According to the project requirements, your app should load all data from json files on startup of the app (books.json, people.json, and rentals.json) However I'm getting no book and no people when
1 - Listing all books or 
2 - Listing all people, See the screenshot below 👇🏻 


image

File: data_controller.rb [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+      JSON.parse(File.read(file)).map(&block)
+    else
+      []
+    end
+  end
+
+  def save_data(file, data)
+    FileUtils.mkdir_p(DATA_DIR)
+    File.write(file, JSON.generate(data))
+  end
```
Comment (BenMukebo) [inline data_controller.rb]: - [ ] Kindly As requested by the previous reviouwer 👇🏻 , Make sure all the data is saved (all the data: books, people, and rentals) into JSON files

> 
>  - [ ] The app works well, I was able to create a person, book, and rental, but looking at the requirements for this project, data wasn't saved. After exiting the app using option 7, data should be saved and then loaded when the app is started. I have attached an image for your reference. You may as well fix the error that comes when a user use option 7 (see attached image).
> 
> save
> 
> _Originally posted by @elisha2kyakpo1 in https://github.com/tanveerisonline/Oop_School_Library/pull/6#discussion_r1138527855_
>

Comment (BenMukebo) [inline data_controller.rb]: - [ ] According to the project requirement 👇🏻 Kindly ensure you properly handle the scenario where the storage files are created and data stored if they do not exist. Actually, the data is no longer saved when I delete one of the JSON files (the app should work and show no errors).

**Review by BenMukebo — CHANGES_REQUESTED** [review]
### Hi @TEAM,

You did a great job so far 👏🏼 👏🏼
There are some issues that you still need to work on to go to the next project but you are almost there!
<FEEL FREE TO CHANGE THIS MESSAGE TO YOUR PERSONAL WORDS>

### To highlight 💪🏻
- No linter errors. ✔️ 
- Gitflow was correctly used.  ✔️ 
 👍  Good job!


## STATUS: CHANGES REQUESTED  ♻️ 
_Please Check the comments under the review and make all required changes_

## Optional suggestions
_Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better._

Cheers and Happy coding!👏👏👏 💻  
Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please ping me @BenMukebo when you comment so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**


------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Review by jasonfelice — APPROVED** [review]
Hi @tanveerisonline, 

Your project is complete!  There is nothing else to say other than... it's time to merge it :shipit: 
Congratulations! 🎉 

<p align="center"></p>

## Highlights
No Linter errors✔️

## Optional suggestions

_Every comment with the [OPTIONAL] prefix won't stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better. Some of them were simply missed by the previous reviewer and addressing them will really improve your application._

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Pull request merged**